web/sharings/drives.go

```diff
@@ -54,41 +55,66 @@ func HeadDirOrFile(c echo.Context, inst *instance.Instance, s *sharing.Sharing)
 		err = middlewares.AllowVFS(c, permission.GET, file)
 	}
 	if err != nil {
-		return err
+		return nil, nil, files.WrapVfsError(err)
 	}
-	return nil
+
+	if dir != nil {
+		return dir, nil, nil
+	}
+	return nil, file, nil
 }
 
-func GetDirOrFile(c echo.Context, inst *instance.Instance, s *sharing.Sharing) error {
-	dir, file, err := inst.VFS().DirOrFileByID(c.Param("file-id"))
-	if err != nil {
-		return files.WrapVfsError(err)
+// Same as `loadDirOrFile` but intolerant of files, responds 422s
+func loadDirFromParam(c echo.Context, inst *instance.Instance, s *sharing.Sharing) (*vfs.DirDoc, error) {
+	dir, file, err := loadDirOrFileFromParam(c, inst, s)
+	if file != nil {
+		return nil, jsonapi.InvalidParameter("file-id", errors.New("file-id: not a directory"))
 	}
+	return dir, err
+}
 
+// Same as `loadDirOrFile` but intolerant of directories, responds 422s
+func loadFileFromParam(c echo.Context, inst *instance.Instance, s *sharing.Sharing) (*vfs.FileDoc, error) {
+	dir, file, err := loadDirOrFileFromParam(c, inst, s)
 	if dir != nil {
-		err = middlewares.AllowVFS(c, permission.GET, dir)
-	} else {
-		err = middlewares.AllowVFS(c, permission.GET, file)
+		return nil, jsonapi.InvalidParameter("file-id", errors.New("file-id: not a file"))
 	}
+	return file, err
+}
+
+func HeadDirOrFile(c echo.Context, inst *instance.Instance, s *sharing.Sharing) error {
+	_, _, err := loadDirOrFileFromParam(c, inst, s)
 	if err != nil {
 		return err
 	}
+	return nil
+}
 
+func GetDirOrFileData(c echo.Context, inst *instance.Instance, s *sharing.Sharing) error {
+	dir, file, err := loadDirOrFileFromParam(c, inst, s)
+	if err != nil {
+		return err
+	}
 	if dir != nil {
 		return files.DirData(c, http.StatusOK, dir)
 	}
 	return files.FileData(c, http.StatusOK, file, true, nil)
 }
 
-func GetDirSize(c echo.Context, inst *instance.Instance, s *sharing.Sharing) error {
-	fs := inst.VFS()
-	fileID := c.Param("file-id")
-	dir, err := fs.DirByID(fileID)
+func ReadFileContentFromIDHandler(c echo.Context, inst *instance.Instance, s *sharing.Sharing) error {
+	//TODO: CSP ?
```

Comment:
Yes, I think we need to add CSP stuff

Reply:
We should test it, but there are also the default CSP rules applied with the secure middleware which may contain the organization instance host, instead of the host of the proxy instance. But, yes, with `frame-ancestors: *`, there are good chances that it works as is.